Adminbar: hide in WP Mobile Apps

enejb opened the pull request:
Related Post p9oQ9f-ow-p2

When request any part of the calypso from the mobile apps (iOS and Android) we want to hide the adminbar both when the user is logged in as well as logged out. 

This done so that the user doesn't see the admin bar navigation inside the app which can cause confusion. Since there is navigation already provided by the Apps themself. 

#### Changes proposed in this Pull Request

* This PR tried to fix the https://github.com/wordpress-mobile/WordPress-Android/issues/13880, https://github.com/wordpress-mobile/WordPress-iOS/issues/7716 issue by making sure that we hide the admin bar view on app views.

#### Testing instructions
* Install an extension to your browser that modifies the headers. 
1. Add a new device in the Resonsive design testing tool. 

<img width="383" alt="Screen Shot 2021-01-25 at 5 19 55 PM" src="https://user-images.githubusercontent.com/115071/105786337-b9ca2100-5f31-11eb-84bd-af1452a6f70f.png">

<img width="807" alt="Screen Shot 2021-01-25 at 5 22 33 PM" src="https://user-images.githubusercontent.com/115071/105786430-ee3ddd00-5f31-11eb-83a6-67be2bcf5e01.png">

the device should have the following agent. 
```
Mozilla/5.0 (Linux; Android 6.0; Android SDK built for x86_64 Build/MASTER; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/44.0.2403.119 Mobile Safari/537.36 wp-android/4.7
```
or 
```
Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/16B91 wp-iphone/12.1
```

* Select the device. Notice that the masterbar as well as the inline help is gone in calypso.
<img width="928" alt="Screen Shot 2021-01-25 at 5 23 10 PM" src="https://user-images.githubusercontent.com/115071/105786486-044b9d80-5f32-11eb-81d5-49f8eca094c1.png">

Page in question that we want to clean up is 
http://calypso.localhost:3000/theme/spearhead/example.com

@tyxla would love your 2 cents on the approach.

enejb committed:
Masterbar: hide on mobile

sarayourfriend commented:
Is there a reason we can't modify the `Masterbar` component directly to render `null` when `isMobileApp` is `true`?

enejb committed:
Use the useragent source provided

enejb committed:
Use the useragent source provided

enejb committed:
Clean up

enejb committed:
Remove the masterbar from the frontend as well.

enejb committed:
Update sidebar style to remove the masterbar gap

enejb committed:
Remove the inline help icon from the mobile apps as well

enejb commented:
The reason why I think we need to modify it on the server side render is because we never want it to show up. 
It is one of the first things that shows up when rendering. 

I tried to clean it up via both the server side as well as via the component. This way we will not try to load the notifications and other things that come along with the masterbar.

enejb commented:
@saramarcondes after some more investigation. I think it would be possible but by having the class there it is easier to modify the other part of calypso and make the adjustments.

sarayourfriend commented:
@emejb It may be slightly easier that way, but could we do both? I think it would be wise to modify [`renderMasterbar`](https://github.com/automattic/wp-calypso/blob/ad4108693463b8ebeda931a2da5577627bba11cd/client/layout/index.jsx#L136) to not even render a masterbar. There may also be a number of other things in `Layout` that are asynchronously loaded you could cut out that are never used in the mobile app's web views and save on load time.

In any case, I'd advocate for adding the classname as well as preventing the masterbar from rendering altogether.

tyxla commented on client/layout/masterbar/style.scss:
```diff
@@ -1,5 +1,21 @@
+	}
+
+	.layout__content {
+		// Adjust the padding as we no longer
```
Let's finish this comment.

enejb replied:
I removed these changes.

tyxla commented on client/layout/masterbar/masterbar.jsx:
```diff
@@ -5,16 +5,23 @@ import classNames from 'classnames';
-	</header>
-);
+const Masterbar = ( { children, className } ) => {
+	return isWpMobileApp() ? null : (
```
This doesn't seem like the right place for this logic - it is too late, and even if we're hiding the masterbar, all the code is loaded anyway. 

Have we considered doing it as early as possible? Ideally, we wouldn't have to load any of the chunks that are coming from components that are related to (embedded in) the masterbar at all.

enejb replied:
This should no be resoled I removed this changes.

tyxla commented on client/blocks/inline-help/index.jsx:
```diff
@@ -136,6 +137,11 @@ class InlineHelp extends Component {
 			return null;
 		}
+
+		if ( isWpMobileApp() ) {
```
Should this be implemented in the `isInlineHelpVisible()` selector so it would be grabbed by the above selector check?

Or even better, can we avoid loading this block and all the related code by asynchronously loading it only if this isn't the mobile app?

enejb replied:
This should be resolved no changes are now needed to this file

tyxla reviewed (commented):
This generally tests well, @enejb. 

However, I share the same thoughts @saramarcondes already expressed. If we're preventing a component from rendering completely, we should fully prevent it from loading its code, instead of loading it and not rendering it. That applies even more strongly when it's about the masterbar, which loads a bunch of code on its own. Following that strategy will positively affect the mobile app, too, because it will cut some bytes down and might make general loading faster for end users.

jsnajdr reviewed (commented):
Most of the functionality for hiding the masterbar and the inline help is already there in the `client/layout` component. You only need to add one more check to the existing `masterbarIsHidden` and `shouldShowInlineHelp` conditions. In both logged-in and logged-out versions of the `<Layout>`.

No other changes, like modifying the `Masterbar` and `InlineHelp` components, should be necessary.

The CSS patch should be more generic, like:
```css
.layout.has-no-masterbar {
  --masterbar-height: 0;

  .layout__content {
    padding-top: 0;
  }
}
```

enejb committed:
Revert the changes to inline-help and masterbar componentes

enejb committed:
Update the changes to layout to remove the masterbar and inline help

enejb committed:
Remove not needed space

enejb committed:
Remove not needed space

enejb commented:
@saramarcondes, @tyxla and @jsnajdr 
Thanks for all the suggestions. ❤️ 

1. I updated the patch to make the changes on the layouts (logged-out and logged-in) instead of the components directly. 
2. I also updated the CSS so on the layout so that look fine when the admin bar is not there. (no gap in the sidebar) and the padding of main column. 
3. I kept the server side rendering class on the body. Mostly just in case there are cases where we missed to remove the masterbar removal. (Should I remove these?) 


Thanks again for all your help.

tyxla reviewed (commented):
Code looks good to me and appears to be working well now.

There are some failing tests that need fixing, though.

enejb committed:
Check for undifined or null

This is done so that they node renderToString() side works as expected.

enejb committed:
Minor fixes

tyxla commented on client/lib/mobile-app/index.js:
```diff
@@ -1,8 +1,14 @@
+/**
+ * @jest-environment jsdom
+ */
```
```suggestion
```

We can remove this; we would need it in the corresponding test file if you wanted to have `window` declared, but since we've double-checked with the `typeof` check below, that won't be necessary either.

tyxla committed:
Remove unnecessary jest-environment in library code

tyxla commented on client/document/index.jsx:
```diff
@@ -122,6 +122,7 @@ class Document extends React.Component {
 						[ 'is-group-' + sectionGroup ]: sectionGroup,
 						[ 'is-section-' + sectionName ]: sectionName,
 						'is-white-signup': sectionName === 'signup',
+						'is-mobile-app-view': app.isMobileApp,
```
In theory, it's possible that `context.app` won't be defined; should we tighten this up so it doesn't blow just in case?

enejb replied:
I think it would always be defined mostly since we use the app values in other parts of calypso. 
But I can add a check for it.

tyxla reviewed (commented):
This is generally looking good, I just have one suggestion.

@jsnajdr what do you think?

enejb committed:
Add check for app

jsnajdr commented on client/document/index.jsx:
```diff
@@ -122,7 +122,7 @@ class Document extends React.Component {
 						[ 'is-section-' + sectionName ]: sectionName,
 						'is-white-signup': sectionName === 'signup',
-						'is-mobile-app-view': app.isMobileApp,
+						'is-mobile-app-view': app && app.isWpMobileApp,
```
Nit: we have optional chaining now 🙂 — `app?.isWpMobileApp`

tyxla replied:
I second this. FWIW, what Jarda suggested will safely return `undefined` if `app` is not defined, which as a falsy value will just be ignored by `classNames()`.

enejb replied:
Fixed

jsnajdr commented on client/layout/index.jsx:
```diff
@@ -134,6 +139,9 @@ class Layout extends Component {
 	renderMasterbar() {
+		if ( this.props.masterbarIsHidden ) {
+			return null;
+		}
```
Could we continue to render the masterbar and hide it with CSS here? Instead of not rendering it at all?

I'm afraid that not rendering it at all could break some flows. For example, some code may rely on `document.getElementById( 'header' )` returning a valid element. I recently fixed one such bug in https://github.com/Automattic/wp-calypso/pull/48046.

Various sections often tweak the layout in rather messy ways, that's why I think this change is risky.

tyxla replied:
The general rationale behind not rendering it was to not load unnecessarily all of the additional components that come together with the masterbar if they will be loaded in vain. To prevent such issues perhaps we could render a placeholder masterbar with the `id="header"`?

jsnajdr replied:
I see, then some `<EmptyMasterbar />` component that just renders the element would be good.

enejb replied:
Thanks for the suggestion I added an EmptyMasterbar component. Let me know what you think?

jsnajdr reviewed (commented):
By the way, there's one more thing that should be hidden in mobile views like this: the cookie banner.

jsnajdr commented:
There is also #21408 from two years ago where @seear did a very similar thing: hide the masterbar on some Jetpack Connect mobile flows. It would be nice to eventually unify the approach. And also avoid rendering the masterbar instead of just hiding it with CSS, after the detailed testing it requires.

enejb committed:
Fix conditional

enejb committed:
Add EmptyMasterbar component

This component is intentionally left empty. So that code doesn't explode and cause error when someone tries do a request for the masterbar.

enejb commented on client/layout/logged-out.jsx:
```diff
@@ -91,7 +92,7 @@ const LayoutLoggedOut = ( {
 		}
-	} else if ( config.isEnabled( 'jetpack-cloud' ) ) {
+	} else if ( config.isEnabled( 'jetpack-cloud' ) || isWpMobileApp() ) {
 		masterbar = null;
```
Should we also return it here?

tyxla replied:
That would make sense to me.

tyxla commented on client/layout/masterbar/empty.jsx:
```diff
@@ -0,0 +1,20 @@
+ */
+
+const EmptyMasterbar = ( { className } ) => (
+	<header id="header" className={ classNames( 'masterbar', className ) } />
```
I think the `className` prop is unnecessary - [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it). If we don't use it, let's ditch it.

enejb replied:
Good call. Removed.

tyxla reviewed (commented):
Other than always returning the empty masterbar component vs `null` sometimes and removing the unnecessary `className` overhead, this looks good to me. Anything to add @jsnajdr?

jsnajdr reviewed (approved):
Looks good :+1: Thanks for addressing all feedback 🙂 

@tyxla I think that rendering nothing instead of `EmptyMasterbar` in logged-out layout should be OK. Logged-out layout has far fewer usages than the logged-in one, and the usages are much simpler.

tyxla reviewed (approved):
This looks good to me too, before with or without addressing the `className` prop nit. 

Thanks for the patience and hard work here, Enej 🍌

enejb committed:
Update the emptyMasterbar to be more YAGNI

enejb committed:
Fix export

dzver commented:
<img width="1313" alt="Screenshot 2021-02-03 at 23 29 59" src="https://user-images.githubusercontent.com/82778/106812109-ea1f6700-6677-11eb-8f53-cf81ccb29225.png">

This breaks the styling of the onboarding and the similar signup flows.
